network/testing.go

```diff
@@ -20,117 +21,250 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-var initialPort = uint64(2000)
+// JoinAndWait is a helper method for joining a destination server
+// and waiting for the connection to be successful (destination node is a peer of source)
+func JoinAndWait(
+	source,
+	destination *Server,
+	connectTimeout time.Duration,
+	joinTimeout time.Duration,
+) error {
+	if joinTimeout == 0 {
+		joinTimeout = DefaultJoinTimeout
+	}
 
-func CreateServer(t *testing.T, callback func(c *Config)) *Server {
-	// create the server
-	cfg := DefaultConfig()
-	cfg.Addr.Port = int(atomic.AddUint64(&initialPort, 1))
-	cfg.Chain = &chain.Chain{
-		Params: &chain.Params{
-			ChainID: 1,
-		},
+	if connectTimeout < joinTimeout {
+		// In case the connect timeout is smaller than the join timeout, align them
+		connectTimeout = joinTimeout
 	}
 
-	logger := hclog.NewNullLogger()
+	// The join routine should be separate
+	go func() {
+		if joinErr := source.Join(destination.AddrInfo(), joinTimeout); joinErr != nil {
+			source.logger.Error(fmt.Sprintf("Unable to join peer, %v", joinErr))
```

Comment:
Why don't we use `t.Error` or `t.Logger` instead of source.logger to show this message?

Reply:
I don't want to involve the `*testing.T` in the context of these helper methods.

If the join fails, the connect will time out nonetheless in the same func

What do you think?

Reply:
But this function is intended to be called with testing Server whose logger is NullLogger as default. This log will be shown?

Reply:
You're right 👍 

I've imagined this function as a simple helper, whose source of truth about connection should come not from the go routine that starts the join process, but from the `WaitUntilPeerConnectsTo`. Sometimes we expect this join to fail, and in that case we don't need the entire test to error out or to clutter the testing logs.

I've removed the error handling in the go routine since it's irrelevant:

4ffc1da